nmigen_soc/wishbone/bus.py

```diff
@@ -166,6 +167,147 @@ def memory_map(self, memory_map):
+        By default the ``sync`` domain is used.
+    """
+    def __init__(self, *, intr_bus, sub_bus, domain="sync"):
+        if not isinstance(intr_bus, Interface):
+            raise TypeError("Initiator bus must be an instance of wishbone.Interface, not {!r}"
+                            .format(intr_bus))
+        if not isinstance(sub_bus, Interface):
+            raise TypeError("Subordinate bus must be an instance of wishbone.Interface, not {!r}"
+                            .format(sub_bus))
+        intr_size = (2**intr_bus.addr_width)*intr_bus.data_width
+        sub_size = (2**sub_bus.addr_width)*sub_bus.data_width
+        if intr_size != sub_size:
+            raise ValueError("Total bit size of initiator and subordinate bus have to be the same")
+        if sub_bus.granularity > intr_bus.granularity:
+            raise ValueError(
+                "Granularity of subordinate bus has to be smaller or equal to "
+                "granulariy of initiator bus"
+            )
+        for opt_output in {"lock"}:
+            if hasattr(intr_bus, opt_output) and not hasattr(sub_bus, opt_output):
+                raise ValueError("Initiator bus has optional output {!r}, but the suborbdinate bus "
+                                 "does not have a corresponding input"
+                                 .format(opt_output))
+        for opt_output in {"err", "rty"}:
+            if hasattr(sub_bus, opt_output) and not hasattr(intr_bus, opt_output):
+                raise ValueError("Subordinate bus has optional output {!r}, but the initiator bus "
+                                 "does not have a corresponding input"
+                                 .format(opt_output))
+        if intr_bus.data_width > sub_bus.data_width:
+            raise NotImplementedError(
+                "Support for multi-cycle bus operation when initiator data_width is"
+                "bigger than the subordinate one is not implemented."
+            )
+
+        self.intr_bus = intr_bus
+        self.sub_bus = sub_bus
+        self.domain = domain
+
+    def elaborate(self, platform):
+        intr_bus = self.intr_bus
+        sub_bus = self.sub_bus
+
+        m = Module()
+
+        common_addr_width = min(intr_bus.addr_width, sub_bus.addr_width)
+        m.d.comb += [
+            sub_bus.cyc.eq(intr_bus.cyc),
+            sub_bus.we.eq(intr_bus.we),
+            sub_bus.adr[(sub_bus.addr_width-common_addr_width):sub_bus.addr_width].eq(
+                intr_bus.adr[(intr_bus.addr_width-common_addr_width):intr_bus.addr_width]
+            ),
+            intr_bus.ack.eq(sub_bus.ack),
+        ]
+        if hasattr(intr_bus, "err"):
+            m.d.comb += intr_bus.err.eq(getattr(sub_bus, "err", 0))
+        if hasattr(intr_bus, "rty"):
+            m.d.comb += intr_bus.rty.eq(getattr(sub_bus, "rty", 0))
+        if hasattr(sub_bus, "lock"):
+            m.d.comb += sub_bus.lock.eq(getattr(intr_bus, "lock", 0))
```

Comment:
The fallback value for LOCK should be 1 on the subordinate side.

Reply:
What is reasoning behind this default ?
It's common in buses of CPUs that they assume bus transactions may be arbitrated unless special signaling is done like a RMW signal or so.
Also the MOS6502 does not have a signal to indicate when there is memory cycle so CYC and STB are always high. If one then defaults to lock=1 this bus will monopolize a downchain arbiter.

Reply:
Sorry for omitting the rationale about this, let's elaborate.

The WB4 specification doesn't (AFAIK) mention whether the bus transactions of an initiator without a LOCK signal are interruptible or not. Section 3.3 does mention uninterruptible access in the context of BLOCK cycles, in a case where *both* initiator and subordinate interfaces have a LOCK signal.

If an initiator doesn't provide a LOCK signal, and 0 was the fallback, then its bus transfers (e.g. bursts) would become interruptible, and it may or may not have been designed with this assumption in mind.

Recommendation 3.05 acknowledges that many arbiters (including ours by default) switch ownership when CYC is low:
> Arbitration logic often uses [CYC_I] to select between MASTER interfaces. Keeping
> [CYC_O] asserted may lead to arbitration problems. It is therefore recommended that
> [CYC_O] is not indefinitely asserted.

A LOCK fallback value of 1 would emulate this behavior, which I believe is safer.

If your CPU relies on keeping CYC permanently asserted, I think it should have a LOCK signal on its own initiator interface and keep it at 0. It may only be compatible with arbiters implementing LOCK anyway.

Reply:
On further thought, I think you're right on this, and that we should default to 0.

Recommendation 3.05 also implies that some arbiters may use something else than CYC to switch bus ownership, without mentioning LOCK. This means that your CPU should be able to work without requiring it on its initiator interface.

Reply:
Will make separate PR for changing the default in Arbiter so we can continue discussion there.
Both things are true:
* It is not recommended to keep CYC high all the time due to how some arbiters work
* Initiator need to use lock signal if they want to guarantee not to be switched off mid cycle.
Anyway I noticed I don't let an initiator with the lock feature connect to subordinate without a lock, which I think should be possible. Otherwise an initiator with a lock always 0 (as suggested above) will not be able to connect to subordinate (f.ex. memory block) without lock signal. Disadvantage is that then the Connector could (accidently) remove the guarantee of not bus switching with lock high.

Reply:
See also #22